### ISSUE TITLE
es-module friendly entry provide

### ISSUE THREAD

**Issue by chigix** [ai/nanoid#128]
Basically, `require` is not available in browser, for which the js source code files are unable to use directly in modern Browser HTML context where ECMA Modules Feature has been supported natively.

I've also noticed that you are using `parcel` to build up a es style module file for usage in demo html file.

Why not directly make it available through es-module-aware entry in `package.json`:
https://github.com/rollup/rollup/wiki/pkg.module

**Comment by ai** [comment]
I am waiting for some real use cases of using ESM.

Right now we do not have tools and best practices. Most of ESM compatible software do it against spec (for instance `import lib from 'lib'` is invalid ESM syntax since you need full path and `.js` on the end). So most of the current implementation is just hype based software.

Right now we have only one ESM compatible bundler (webpack and Parcel anyway convert ESM to `require` hack) pika. But this bundler doesn’t solve round-trip latency problem and as a result, have webpage will have a bad performance.

I am not against ESM, we will find ways and it is a great technology. But we should use technology with understanding how to use it correctly, not for the hype.

**Comment by ai** [comment]
If you have any real use case for ESM, I can reconsider my opinion.

Right now I planned to investigate ESM best practices at the end of the summer.

**Comment by TrySound** [comment]
The only ESM compatible bundler is Rollup. Pika is just a wrapper around it.

**Comment by ai** [comment]
@TrySound I mean who have real production solution (Babel can parse ESM keywords too). We need to solve the round-trip latency problem. We need to support old `require()` in legacy dependencies.

**Comment by chigix** [comment]
Briefly, I would recommend the [rxjs' package.json](https://cdn.jsdelivr.net/npm/rxjs/package.json) practice, where scenarios from node platform to webpackage usage in browser and tree shaking friendly practice are covered for users.

Initially, it is impossible to make one code work as common js module and ES style module. Hence, building is widely employed by many npm packages supporting node and browser platform, but would also be the main challenge for nanoid, I think.

But `nanoid` aims at extremely small and simple, for which I am imagining that it might be acceptable to make up es module style file from current common js module code through grunt directly as distribution. Personally, it would be nice to employ javascript compilation like typescript or babel though.

**Comment by ai** [comment]
> Briefly, I would recommend the rxjs' package.json practice, where scenarios from node platform to webpackage usage in browser and tree shaking friendly practice are covered for users.

Can you show some production system which use ESM of rxjs? (without converting modules to requires like Webpack does)

> Initially, it is impossible to make one code work as common js module and ES style module.

Another problem is that `package.module` will work only for main entry `index.js`.

How we will work with `import nanoid from 'nanoid/async'` and `require('nanoid/async')`? Node.js has a proposal to put `nanoid/async/index.mjs` for ESM and `nanoid/async/index.js` for CJS. But this proposal was not yet approved.

> Personally, it would be nice to employ javascript compilation like typescript or babel though.

Avoiding Babel was the original reason why Nano ID is so small. Babel adds a lot of unnecessary stuff.

> But nanoid aims at extremely small and simple, for which I am imagining that it might be acceptable to make up es module style file from current common js module code through grunt directly as distribution. 

Yeap, when we will have real production ready ESM system, I will make simple regexps to convert ESM to CJS.

**Comment by TrySound** [comment]
> How we will work with import nanoid from 'nanoid/async' and require('nanoid/async')? Node.js has a proposal to put nanoid/async/index.mjs for ESM and nanoid/async/index.js for CJS. But this proposal was not yet approved.

Here's one of approaches for path imports.
https://unpkg.com/react-transition-group@4.1.1/Transition/package.json

**Comment by chigix** [comment]
It seems react transition has employed the `module` entry for indicating es module index file path as well.

**Comment by TrySound** [comment]
Yes, both cases are supported.

**Comment by chigix** [comment]
@ai Because rxjs' code base is very huge, for which rxjs relies on tree shaking in practice, where esm files are loaded by browsers through "import" after webpack built and bundled. The example of this practice has been employed by Google Angular. You could directly experience and check this action through a simple getting started with Angular: https://stackblitz.com/edit/angular-qvrcjd

However, rxjs provides an all-in-one umd distribution file for browsers loading the module directly without any third party loader: https://jsfiddle.net/1g3v57kc/37/

**Comment by Justineo** [comment]
ESM is useful for bundlers (webpack/Rollup) to perform static analysis and treeshaking, whether they use “`require` hack” or not, which finally leads to smaller bundle sizes. I wouldn't say it's purely for the hype.

**Comment by ai** [comment]
@Justineo how exactly webpack or Rollup can optimize nanoid wirh ESM compare to CommonJS?

**Comment by TrySound** [comment]
Webpack wraps every commonjs module with function. Check this output
webpack.config.js
```js
module.exports = {
  mode: 'production'
}
```
src/index.js
```js
import nanoid from 'nanoid';
console.log(nanoid)
```
cjs.js contains `module.exports = function ...`

esm.js contains `export default function ...`

https://gist.github.com/TrySound/374824e966fbe3574f77cda1082c5ef4

**Comment by TrySound** [comment]
When project has many commonjs modules bundle can't use scope hoisting and minifier can't optimise many things.

**Comment by ai** [comment]
@TrySound can you compare webpack output with [ESM version of Nano ID](https://www.npmjs.com/package/nanoid-esm)?

**Comment by TrySound** [comment]
This is it. I just manually changed module.exports to export default in node_modules.
esm - 1165b
cjs - 1214b

**Comment by TrySound** [comment]
nanoid-esm has very weird esm support. I'm not surprised the size is bigger than commonjs. Yes, you get scope hoisting but minifier will probably not be able to inline the function when it's necessary.
https://unpkg.com/browse/nanoid-esm@2.1.0-esm.4/dist-esm/index.browser.js

**Comment by ai** [comment]
@TrySound good research

Does webpack support Node.js’s [exports map](https://github.com/jkrems/proposal-pkg-exports)? (I need it for `import async from "nanoid/async"`)

**Comment by TrySound** [comment]
Nope. I think nothing will support this proposal until it released as stable. module field is minimal viable way for now. mjs is not reliable in many cases. Also typescript users can be angry because of module.exports instead of `exports.default = ` (nobody want to enable esModuleInterop).

**Comment by TrySound** [comment]
Multiple entry points can be implemented with dumb package.jsons
https://unpkg.com/browse/date-fns@2.6.0/addDays/package.json

**Comment by ai** [comment]
OK, I prefer to wait for removeing flag for ESM in Node.js 13

**Comment by chase-moskal** [comment]
ok i spent way too long writing too much comment here, but hopefully somebody will gain something here ;)

> OK, I prefer to wait for removeing flag for ESM in Node.js 13

nanoid seems focused on exclusively supporting the traditional node/cjs ecosystem

over the last year, i've been treating cjs as a legacy format which i only use in node environments, and my new browser frontend apps are *esm-only.* working with the new ecosystem is really nice, and all of the modern packages i care about are supporting esm-first nowadays

from your perspective, i wonder if it might seem like esm is a wacky newfangled "experimental" or "unstable" workflow (perhaps mostly because node has been dragging its feet quite pathetically on the esm front). however, from my perspective, esm is a bustling ecosystem full of new ideas, rapid development, idealized workflows, the best new packages, established best practices, and is the obvious future of web development generally

unfortunately, today, devs like myself participating in the new paradigm are *not able to import nanoid into their apps.* this makes nanoid only a "node" library, but unfortunately not suitable as a "browser" or "isomorphic" library

> nanoid-esm has very weird esm support.

yes, nanoid-esm was just a disgusting babel transform, and some tweaks to remove some node-environment-specific things

nanoid is one of a few commonjs stragglers i've encountered over the last year. since my nanoid-esm fork is hacky, bloated, and unmaintained, i ultimately decided to replace nanoid in my application with merely

    export function generateId(n = 1) {
      n -= 1
      return Math.random().toString(36).substring(2)
        + (n > 0 ? generateId(n) : "")
    }

surely, this is tragically inferior to using nanoid, and in more ways than i'm aware. however, it *seems* sufficient for my currently-limited use-case, and provides a simple way for me to generate id's clientside, without worrying about maintaining a hacky fork or convincing you guys to embrace this emerging ecosystem

hopefully in the future, when you folks feel more comfortable that esm isn't just a fad, i can eventually switch back to using nanoid for cases like this (hopefully a fully environment-agnostic isomorphic edition of nanoid that works natively in browsers, node, and deno -- that's my feature request!)

i think a good hypothetical path to upgrading nanoid to esm-first could look like this:
1. convert source code to typescript
    - generate esm in `dist/` and set "module" to `dist/index.js`
    - generate cjs in `dist-cjs/` and set "main" to `dist-cjs/index.js`
    - generate type declaration `.d.ts` files in both locations
2. eliminate all environment-specific code, so the library can be truly isomorphic
    - test for browser environment
    - test for node environment
    - test for deno environment

food for thought. and of course, i mean no disrespect at all here, i'm just hoping to shed light and rationale on a different perspective and bounce some ideas around

nanoid is great work, i hope i can contribute more in the future so that more devs can use it everyday. cheers!

:wave: chase

**Comment by ai** [comment]
Node.js 13.2 was released with unflagged ESM support and (as promised) I am starting to work on ESM-version of Nano ID.

I am planning to use [`package.exports`](https://nodejs.org/api/esm.html#esm_package_exports) with `.mjs` file ext and small script to build CJS files from ESM files.

**Comment by TrySound** [comment]
@ai Why do you want mjs extension? Afaik package.exports works with .js too. .mjs will still confuse webpack before they will make it closer to spec.

**Comment by ai** [comment]
@TrySound what is the problem with `.mjs` exactly?

**Comment by TrySound** [comment]
All .js files with with esm cannot import .mjs.

Extensions are weird. TS modules with jsx could be .mtsx. Just an opinion.

**Comment by ai** [comment]
I post issue to `clean-publish` to add compiling step: https://github.com/shashkovdanil/clean-publish/issues/39

> All .js files with with esm cannot import .mjs. Event if mjs is specified explicitly in import path.

Good point. OK, let’s use `.js`.

> Afaik package.exports works with .js too. .mjs will still confuse webpack before they will make it closer to spec.

BTW, webpack doesn’t support `package.exports` right now. But anyway current webpack’s `package.browser` doesn’t support ESM, so any way we need a webpack to implement  `package.exports`.

**Comment by ai** [comment]
I updated my [ESLint config](https://github.com/logux/eslint-config) to support `import` and tested real Node.js modules of [logux.io](https://github.com/logux/logux.io) (it is sad that even `__dirname` is not avaiable without dirty code and `import.meta` is not compatible with ESLint).

**Comment by ai** [comment]
I created a tool for myself to publish Nano ID as ESM
https://github.com/ai/dual-publish/

I will test it on NanoEvents and will start preparing next major release for Nano ID.

**Comment by ai** [comment]
I released [Nano Events](https://github.com/ai/nanoevents/) as an experiment of my new dual ESM/CJS system.

Does everyone like how it work? Nano ID will use the same system.

**Comment by ai** [comment]
We found a big problem during our migration to ESM. Webpack and Node.js have different behavior for default export for dual ESM/CJS packages which make default export unusable:

```js
// lib/index.cjs
module.exports = 'a'

// lib/index.mjs
export default 'b'

// Node.js
import lib from 'lib' //=> 'a'
require('lib') //=> 'b'

// Webpack
import lib from 'lib' //=> 'a'
require('lib') //=> {default: 'a' }
```

https://github.com/storeon/storeon/issues/82a

@TrySound do you have an idea hot to solve it without going to named exports (which will increase the size of nanoid)t?

**Comment by TrySound** [comment]
Only commonjs version will be increased. ESM will be scope hoisted whatever import you use. Not that big deal for node apps.

This is what I meant. Default exports give too much problems. Named exports just work.

**Comment by ai** [comment]
> Default exports give too much problems.

I am afraid that named exports will force C-style of having long prefixes in libraries.

For instance, how Nano ID API should look like?

This one looks too long:

```js
import { nanoid, nanoidAsync, nanoidUnsafe, nanoidGenerate } from 'nanoid'
```

This one looks too complicated:

```js
import { async as nanoid } from 'nanoid'
```

Also, I do not like that Node.js will need to compile and initialize big file with all exports just for a single small function.

**Comment by TrySound** [comment]
Anyway, straightforward fix is to replace `module.exports = fn` with `exports.default = fn`. This way you can even drop tricky namespace and export from ts definitions in favour of just `export default`.

**Comment by ai** [comment]
> Anyway, straightforward fix is to replace module.exports = fn with exports.default = fn.

It will force ugly `require('lib').default` for all Node.js users. Node.js does not have stable ESM yet to think about non-import users as a minority.

**Comment by TrySound** [comment]
One more solution is using named export even with single export. This will require some noise to prevent collisions, though in my experience it's not often case.
```
const { nanoid } = require('nanoid')
const { nanoid: asyncNanoid } = require('nanoid/async')
```
```
import { nanoid } from 'nanoid'
import { nanoid as asyncNanoid } from 'nanoid/async'
```

All solutions have tradeoffs.

**Comment by ai** [comment]
I moved Storeon, Nano Events, and Nano Delay to named exports

https://github.com/storeon/storeon
https://github.com/ai/nanoevents
https://github.com/ai/nanodelay

Do it work correctly?

**Comment by TrySound** [comment]
This is great!

**Comment by ai** [comment]
@TrySound does tree-shaking really?

This change reduced the size of the page from 69 KB to 40 KB
https://github.com/logux/client/commit/8130d4b8bc6692b683682526d1f8df310cfdbf82

Seems like there is a problem in webpack and Parcel with nested tree-shaking: project → lib1 → lib2 (lib2 will not be tree-shaked).

**Comment by TrySound** [comment]
In this particular case I think the problem was in commonjs. Or did you try import syntax too?

**Comment by ai** [comment]
@TrySound yeap, I tried to use `import` and still show 69 KB. Only direct import solved the problem.

**Comment by TrySound** [comment]
How did you test? With commonjs git dependency?
![image](https://user-images.githubusercontent.com/5635476/76849237-16f73f80-6856-11ea-88df-b09d014400e2.png)

**Comment by ai** [comment]
@TrySound got it. I forget that the GitHub version has only CJS exports.

Thanks for double-checking to me.

**Comment by ai** [comment]
The ESM support was released in Nano ID 3.0
